### DISCUSSION TITLE
Peclet number characteristic length

### DISCUSSION THREAD

**Discussion by rebpoli** [libMesh/libmesh#3111] [Q&A]
Hi,

I am trying to stabilize a convection dominated problem. I am trying to calculate the Peclet number and for that I have use for the element characteristic length, defined as the element size in the flow direction.

I understand there are possible simplifications, but I am wondering if libmesh have hooks to help me with that.
I can see functions like Elem::volume, Elem::hmax, Elem::hmin, for example ...

Any idea on how I can do that?

Thanks,
Renato

**Comment by roystgnr** [comment]
There's no API like that currently.

It sounds like it would be useful enough to add, except that I'm not sure how it would be defined in general.  I'd guess "h_x cos(theta) + h_y sin(theta)" for flow at an angle theta to a cartesian quad grid, for example, but what's the definition for flow at an angle to an arbitrarily oriented (or even twisted or quadratic and curved) prism?

**Comment by jwpeterson** [comment]
I copied this formula from somewhere, I think @benkirk's dissertation maybe? 

![flow_aligned](https://user-images.githubusercontent.com/1775907/147986722-7f6f4b2d-9bd4-48fb-aa01-5c71831eff0b.png)

Some care is needed to avoid dividing by zero, and the formula involves the inverse Jacobian for the element map, so I think not convenient to implement as an `Elem` member function, since you need some `FE` data.

**Reply by jwpeterson** [reply]
After double-checking more carefully, looks like this is the one from Ben's dissertation:
![benkirk_flow_aligned](https://user-images.githubusercontent.com/1775907/147989315-1a29ffb8-4edc-4a78-adcc-4b902c586ce4.png)
NN = number of nodes on current Elem
\phi_k = Lagrange basis function k on current Elem 
\hat{u} = unit vector in direction of velocity

**Reply by roystgnr** [reply]
Oh, that's nice.  I remembered Ben using a nice general stabilization parameter formula like that, but I'd forgotten that it wasn't even technically hypersonics-specific, that it ended up giving a single generally-applicable length scale.

**Comment by rebpoli** [comment]
Thanks.

I can add that to the assembly loop.
It looks enough for now.

Best regards, and happy new year!

Renato


Em seg., 3 de jan. de 2022 19:59, John W. Peterson ***@***.***>
escreveu:

**Comment by benkirk** [comment]
What is the PDE, a scalar transport equation or a system?  For scalars, something like above is usually good enough, the idea being define some length scale aligned with the flow direction.  For linear elements anyway you can envision the shape-function dot product thing working out.

For systems like compressible Navier Stokes & MHD specifically, things get more complicated and theres some nice approaches from Glasby & Anderson using eigenvalue transformation magic. I can provide some references if interested.

**Comment by rebpoli** [comment]
Hi Benjamin,

Is is a convection dominated PDE.

We are working with thermal-poro-elasticity, so displacements and pressure
are solved fully coupled while temperature is solved sequentially in a
convection-diffusion PDE. Velocity comes from the pressure gradient
(Darcy's law).

It happens that diffusion is very low and the system oscillates near the
well (at higher velocities).

I am trying to implement a simple Petrov Galerkin scheme, believing it
won't get more complex than that... but we are not sure yet...

Any suggestions and good references are of course really welcome!

Thanks for the help.
Best regards,
Renato

Em ter., 4 de jan. de 2022 14:15, Benjamin S. Kirk ***@***.***>
escreveu:

**Reply by benkirk** [reply]
For a scalar PDE (temperature transport with a given velocity field), finding h as above should hold together.

For coupled systems of equations, with different characteristics that each behave differently, thing can get more complicated.  The most effective stuff I found is summarized below.  Much of this is in the recent AIAA literature.

https://arc.aiaa.org/doi/pdf/10.2514/6.2013-2559
http://crossmark.crossref.org/dialog/?doi=10.2514%2F6.2013-2559&domain=pdf&date_stamp=2013-06-22
 
The paper and presentation are here:
https://drive.google.com/drive/folders/1CoXfw-ASxwlcm6g0IVsBu_V4JERMXgLt?usp=sharing
 
is the probably most relevant summary of where my efforts went asymptotic a number of years ago.
 
I focused almost exclusively on shock dominated & P=1 SUPG.  Using the tau described in here that worked great for hanging node AMR, with typical 2:1 stretching there.
 
Ryan Glasby’s taken compressible Navier Stokes SUPG to the next level, but with less focus on shock dominated flows:
https://arc.aiaa.org/doi/10.2514/6.2016-0567
https://arc.aiaa.org/doi/10.2514/6.2016-1054
https://arc.aiaa.org/doi/pdf/10.2514/6.2018-0781
https://arc.aiaa.org/doi/10.2514/6.2019-1344
 
Kyle Anderson’s recent AIAA SUPG, P=1 & P=2
https://arc.aiaa.org/doi/10.2514/6.2018-1307
https://arc.aiaa.org/doi/10.2514/1.C034482
Kyle's been working SUPG internal to NASA’s FUN3D framework as well, I believe.

**Comment by rebpoli** [comment]
Thank you very much!

Best regards,
Renato

Em ter., 4 de jan. de 2022 às 18:04, Benjamin S. Kirk <
***@***.***> escreveu:
